embedded/store/immustore.go

```diff
@@ -2737,3 +2828,163 @@ func minUint64(a, b uint64) uint64 {
+		the situation where a transaction (n+1) could be inserted
+		in the value log before transaction (n)
+						  discard point
+								|
+								v
+				--------+-------+--------+----------
+						|       |        |
+					tn+1:vx   tn:vx  tn-1:vx
+						|                |
+						+----------------+
+						max concurrency
+							range
+		If the log is truncated upto tn, it could lead to removal of
+		data for tn+1. To avoid this overlap, we first go back from
+		the discard point to fetch offsets across vlogs for safe delete
+		points, and then check for transactions further than the
+		discard point to figure out the least offset from the range
+		which can safely be deleted. This offset, tn+1 in the above
+		scenario is the safest point to discard upto to avoid deletion
+		of values for any future transaction.
+	*/
+
+	s.logger.Infof("running truncation up to transaction '%d'", minTxID)
+
+	var err error
+	// tombstones maintain the minimum offset for each value log file that can be safely deleted.
+	tombstones := make(map[byte]int64)
+
+	readFirstEntryOffset := func(id uint64) (*TxEntry, error) {
+		return s.readTxOffsetAt(id, false, 1)
+	}
+
+	back := func(id uint64) error {
+		first, err := readFirstEntryOffset(id)
+		if err != nil {
+			return err
+		}
+
+		// Iterate over all past transactions and store the minimum offset for each value log file.
+		v, off := decodeOffset(first.VOff())
+		if _, ok := tombstones[v]; !ok {
+			tombstones[v] = off
+		}
+		return nil
+	}
+
+	front := func(id uint64) error {
+		first, err := readFirstEntryOffset(id)
+		if err != nil {
+			return err
+		}
+
+		// Check if any future transaction offset lies before past transaction(s)
+		// If so, then update the offset to the minimum offset for that value log file.
+		v, off := decodeOffset(first.VOff())
+		if val, ok := tombstones[v]; ok {
+			if off < val {
+				tombstones[v] = off
+			}
+		}
+		return nil
+	}
+
+	// Walk back to get offsets across vlogs which can be deleted safely.
+	// This way, we can calculate the minimum offset for each value log file.
+	{
+		var i uint64 = minTxID
+		for i > 0 && len(tombstones) != s.MaxIOConcurrency() {
+			err = back(i)
+			i--
+			if err != nil { // if txn not found, then continue as previous txn could have been deleted
+				s.logger.Errorf("failed to fetch transaction %d {traversal=back, err = %v}", i, err)
+				return err
+			}
+		}
+	}
+
+	// Walk front to check if any future transaction offset lies before past transaction(s)
+	{
+		// Check for transactions upto the last committed transaction to avoid deletion of values for any future transaction.
+		maxTxID := s.LastCommittedTxID()
+		s.logger.Infof("running truncation check between transaction '%d' and '%d'", minTxID, maxTxID)
+
+		// TODO: add more integration tests
+		// Iterate over all future transactions to check if any offset lies before past transaction(s) offset.
+		for j := minTxID; j <= maxTxID; j++ {
+			err = front(j)
+			if err != nil {
```

Comment:
it's not clear to me why `front` will return an error according to the description in the comment

Reply:
If there is any error in reading the transaction, there is some corruption in data etc, I am returning the same error when the value is read from the log

Reply:
And the error is returned to avoid deleting any record accidentally if that transaction can't be read. Could be signs of a corrupted data record

Reply:
ah ok, so it's a validation that all transactions greater than minTxID won't be truncated